homeassistant/components/modbus/cover.py

```diff
@@ -0,0 +1,238 @@
+        # Intermediate states are not supported in such a setup.
+        if self._coil is not None and self._status_register is None:
+            self._state_closed = False
+            self._state_open = True
+            self._state_closing = None
+            self._state_opening = None
+
+        # If we read cover status from the main register (i.e., an optional
+        # status register is not specified), we need to make sure the register_type
+        # is set to "holding".
+        if self._register is not None and self._status_register is None:
+            self._status_register = self._register
+            self._status_register_type = CALL_TYPE_REGISTER_HOLDING
+
+    async def async_added_to_hass(self):
+        """Handle entity which will be added."""
+        state = await self.async_get_last_state()
+        if not state:
+            return
+        self._value = state.state
+
+        async_track_time_interval(
+            self.hass, lambda arg: self.update(), self._scan_interval
+        )
+
+    @property
+    def device_class(self) -> Optional[str]:
+        """Return the device class of the sensor."""
+        return self._device_class
+
+    @property
+    def name(self):
+        """Return the name of the switch."""
+        return self._name
+
+    @property
+    def supported_features(self):
+        """Flag supported features."""
+        return SUPPORT_OPEN | SUPPORT_CLOSE
+
+    @property
+    def available(self) -> bool:
+        """Return True if entity is available."""
+        return self._available
+
+    @property
+    def is_opening(self):
+        """Return if the cover is opening or not."""
+        return self._value == self._state_opening
+
+    @property
+    def is_closing(self):
+        """Return if the cover is closing or not."""
+        return self._value == self._state_closing
+
+    @property
+    def is_closed(self):
+        """Return if the cover is closed or not."""
+        return self._value == self._state_closed
+
+    @property
+    def should_poll(self):
+        """Return True if entity has to be polled for state.
+
+        False if entity pushes its state to HA.
+        """
+
+        # Handle polling directly in this entity
+        return False
+
+    def open_cover(self, **kwargs: Any) -> None:
+        """Open cover."""
+        if self._coil is not None:
+            self._write_coil(True)
+        else:
+            self._write_register(self._state_open)
+
+    def close_cover(self, **kwargs: Any) -> None:
+        """Close cover."""
+        if self._coil is not None:
+            self._write_coil(False)
+        else:
+            self._write_register(self._state_closed)
+
+    def update(self):
+        """Update the state of the cover."""
+        if self._coil is not None and self._status_register is None:
+            self._value = self._read_coil()
+        else:
+            self._value = self._read_status_register()
+
+        self.async_write_ha_state()
```

Comment:
We need to use the sync api in sync context. Use `self.schedule_update_ha_state`.